pkg/component/controller/noderole.go

```diff
@@ -0,0 +1,122 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package controller
+
+import (
+	"context"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/go-openapi/jsonpointer"
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/kubernetes"
+
+	"github.com/k0sproject/k0s/pkg/constant"
+	k8sutil "github.com/k0sproject/k0s/pkg/kubernetes"
+)
+
+// NodeRole implements the component interface to manage node role labels for worker nodes
+type NodeRole struct {
+	kubeClientFactory k8sutil.ClientFactoryInterface
+
+	log     *logrus.Entry
+	k0sVars constant.CfgVars
+}
+
+// NewNodeRole creates new NodeRole reconciler
+func NewNodeRole(k0sVars constant.CfgVars, clientFactory k8sutil.ClientFactoryInterface) (*NodeRole, error) {
+	log := logrus.WithFields(logrus.Fields{"component": "noderole"})
+	return &NodeRole{
+		kubeClientFactory: clientFactory,
+		log:               log,
+		k0sVars:           k0sVars,
+	}, nil
+}
+
+// Init no-op
+func (n *NodeRole) Init() error {
+	return nil
+}
+
+// Run checks and adds labels
+func (n *NodeRole) Run(ctx context.Context) error {
+	client, err := n.kubeClientFactory.GetClient()
+	if err != nil {
+		return err
+	}
+	go func() {
+		timer := time.NewTicker(1 * time.Minute)
+		defer timer.Stop()
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-timer.C:
+				nodes, err := client.CoreV1().Nodes().List(ctx, metav1.ListOptions{})
+				if err != nil {
+					n.log.Errorf("failed to get node list: %v", err)
+					continue
+				}
+
+				for _, node := range nodes.Items {
+					err = n.ensureNodeLabel(ctx, client, node)
+					if err != nil {
+						n.log.Error(err)
+					}
+				}
+			}
+		}
+	}()
+
+	return nil
+}
+
+func (n *NodeRole) ensureNodeLabel(ctx context.Context, client kubernetes.Interface, node corev1.Node) error {
+	var labelToAdd string
+	for label, value := range node.Labels {
+		if strings.HasPrefix(label, constant.NodeRoleLabelNamespace) {
+			return nil
+		}
+
+		if label == constant.K0SNodeRoleLabel {
+			labelToAdd = fmt.Sprintf("%s/%s", constant.NodeRoleLabelNamespace, value)
+		}
+	}
+
+	_, err := n.addNodeLabel(ctx, client, node.Name, labelToAdd, "true")
+	if err != nil {
+		return fmt.Errorf("failed to set label '%s' to node %s: %w", labelToAdd, node.Name, err)
+	}
+	return nil
+}
+
+func (n *NodeRole) addNodeLabel(ctx context.Context, client kubernetes.Interface, node, key, value string) (*corev1.Node, error) {
+	keyPath := fmt.Sprintf("/metadata/labels/%s", jsonpointer.Escape(key))
+	patch := fmt.Sprintf(`[{"op":"add", "path":"%s", "value":"%s" }]`, keyPath, value)
+	return client.CoreV1().Nodes().Patch(ctx, node, types.JSONPatchType, []byte(patch), metav1.PatchOptions{})
```

Comment:
why not to use client.CoreV1().Nodes().Update and pass node instance there insterad of building raw json?

Reply:
My reasoning was that it may cause a race condition: someone or something could change the node in between Get and Update calls.

Looks like it's a real case: https://github.com/kubernetes/client-go/issues/414

Reply:
yeah, it's common thing in go-client, I think it is even expected to retry on already modified objects